### DISCUSSION TITLE
The database pickle

### DISCUSSION THREAD

**Discussion by da2x** [Taiko2k/Tauon#724] [Ideas]
Everything seems to be an unstructured pickled Python object dump.

Any chance you could consider migrating to a an SQLite database file?

It’s difficult for a third-party program to work with the Tauon database without a structured data format.

Currently, changes appear to only be saved on exit when Tauon dumps all its pickles do disk. This may cause dataloss. E.g. if the app crashes after a six hour session without a normal exit. An SQLite database would let you save state changes much more frequently without having to rewrite the entire state file on disk every time.

**Comment by Taiko2k** [comment]
I experimented with SQL early in development and wasn't satisfied with the performance I got. Perhaps that could have been alleviated, but its a bit late now being so far in development (many functions modify data structures directly, its not abstracted). For performance I think I would still want a *read all at once, dump it all later* model anyway.

Tauon currently does have mechanisms to dump while the application is running. Actions such as modifying a playlist should flag it for dumping (it does it once the cursor has left the window). So generally data loss should be minimal. Changing settings don't currently though.

I think the way forward is just tweaking this existing system where issues are found.
